cfavml-complex/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+#[cfg(target_arch = "x86")]
+use core::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use core::arch::x86_64::*;
+
+use cfavml::danger::{Avx2, SimdRegister};
+use num_complex::Complex;
+pub trait ComplexOps<T>
+where
+    Complex<T>: Copy,
+{
+    type Register: Copy;
+    /// Extracts the real component from a register
+    unsafe fn real_values(value: Self::Register) -> Self::Register;
+    /// Extracts the imaginary component from a register
+    unsafe fn imag_values(value: Self::Register) -> Self::Register;
+
+    /// Duplicates the real and imaginary components of a the input into separate registers
+    unsafe fn duplicate_complex_components(
+        value: Self::Register,
+    ) -> (Self::Register, Self::Register) {
+        (Self::real_values(value), Self::imag_values(value))
+    }
+    /// Swaps the real and imaginary components of a complex number
+    unsafe fn swap_complex_components(value: Self::Register) -> Self::Register;
+}
+
+pub struct Avx2Complex;
+impl ComplexOps<f32> for Avx2Complex {
+    type Register = __m256;
+
+    #[inline(always)]
+    unsafe fn real_values(value: Self::Register) -> Self::Register {
+        _mm256_moveldup_ps(value)
+    }
+
+    #[inline(always)]
+    unsafe fn imag_values(value: Self::Register) -> Self::Register {
+        _mm256_movehdup_ps(value)
+    }
+    #[inline(always)]
+    unsafe fn swap_complex_components(value: Self::Register) -> Self::Register {
+        _mm256_permute_ps(value, 0xB1)
+    }
+}
+
+impl SimdRegister<Complex<f32>> for Avx2Complex {
+    type Register = __m256;
+
+    #[inline(always)]
+    unsafe fn load(mem: *const Complex<f32>) -> Self::Register {
+        _mm256_loadu_ps(mem as *const f32)
+    }
+
+    #[inline(always)]
+    unsafe fn filled(value: Complex<f32>) -> Self::Register {
+        // no
+        _mm256_setr_ps(
+            value.re, value.im, value.re, value.im, value.re, value.im, value.re,
+            value.im,
+        )
+    }
+
+    #[inline(always)]
+    unsafe fn zeroed() -> Self::Register {
+        _mm256_setzero_ps()
+    }
+
+    #[inline(always)]
+    unsafe fn add(l1: Self::Register, l2: Self::Register) -> Self::Register {
+        _mm256_add_ps(l1, l2)
+    }
+
+    #[inline(always)]
+    unsafe fn sub(l1: Self::Register, l2: Self::Register) -> Self::Register {
+        _mm256_sub_ps(l1, l2)
+    }
+
+    #[inline(always)]
+    unsafe fn mul(left: Self::Register, right: Self::Register) -> Self::Register {
+        // (a + bi)(c + di) = (ac - bd) + (ad + bc)i
+        // [a, b, a, b, a, b, a, b], [c, d, c, d, c, d, c, d]
+        // [a x 4], [b x 4]
+        let (left_real, left_imag) =
+            <Avx2Complex as ComplexOps<f32>>::duplicate_complex_components(left);
+        let right_shuffled =
+            <Avx2Complex as ComplexOps<f32>>::swap_complex_components(right);
+
+        let output_right = _mm256_mul_ps(left_imag, right_shuffled);
+        _mm256_fmaddsub_ps(left_real, right, output_right)
```

Comment:
So I think although it might be a bit of a PITA is this is likely going to need a `Avx2` and `Avx2Fma` impl, technically within CFAVML, `Avx2` assumes only `avx2` is enabled with no FMA support. Although unlikely, for consistency we should make sure all the crates follow this pattern as well. 

Internally what CFAVML does is on any routine which doesn't actually _need_ FMA, it just calls the `Avx2::example_op`

Reply:
A crate that might be useful is [duplicate](https://docs.rs/duplicate/2.0.0/duplicate/). 

I explored writing a macro rule to implement functions that just called the same function name from another struct, but I couldn't figure out how to pass in the fn call as a raw string to use it in both the definition and the inner call. It started to feel like one of those "If I can, maybe I shouldn't" problems.